File: mobile/src/services/deviceAudit.ts
import {
  loadTechnicianId,
  loadGpsState, storeGpsState,
  loadNetState, storeNetState,
  loadBattOptState, storeBattOptState,
  loadPermState, storePermState, type PermLevel,
  loadLastFixTs,
} from './offlineQueue';
import { reportDeviceEvent } from './sensorService';

// ── Auditoría de dispositivo (evidencia de sabotaje al rastreo) ───────────────
// Detecta TRANSICIONES de estado del teléfono (GPS, red, optimización de batería)
// comparando contra el último estado persistido, y deja un motion_event con hora
// y última posición conocida. Es la pieza que cierra el hueco principal: como el
// estado vive en AsyncStorage, funciona igual desde la UI (sondeo cada 4 s) que
// desde el watchdog/background task (app cerrada), así que apagar el GPS o los
// datos "a escondidas" en segundo plano TAMBIÉN deja huella.
//
// Reglas comunes a todas:
//   - Solo emite si hay una sesión de rastreo activa (technicianId persistido).
//   - El primer muestreo (estado previo == null) solo SIEMBRA el estado, no
//     dispara evento (evita un falso positivo al abrir la app).
//   - Siempre persiste el estado nuevo, haya o no sesión, para tener continuidad.

/** Siembra los estados con la foto actual al iniciar una sesión, sin emitir eventos. */
export async function seedDeviceState(opts: {
  gpsOn?: boolean;
  connected?: boolean;
  batteryOptimized?: boolean;
  perm?: PermLevel;
}): Promise<void> {
  if (opts.gpsOn !== undefined)            await storeGpsState(opts.gpsOn);
  if (opts.connected !== undefined)        await storeNetState(opts.connected);
  if (opts.batteryOptimized !== undefined) await storeBattOptState(opts.batteryOptimized);
  if (opts.perm !== undefined)             await storePermState(opts.perm);
}

/** GPS encendido/apagado. Devuelve el evento emitido (o null si no hubo transición). */
export async function auditGps(servicesOn: boolean): Promise<'gps_on' | 'gps_off' | null> {
  const prev = await loadGpsState();
  await storeGpsState(servicesOn);
  if (prev === null || prev === servicesOn) return null;
  const techId = await loadTechnicianId();
  if (!techId) return null;
  const ev = servicesOn ? 'gps_on' : 'gps_off';
  await reportDeviceEvent(techId, ev);
  return ev;
}

/** Datos/Wi-Fi conectados o no. Devuelve el evento emitido (o null). */
export async function auditNet(connected: boolean): Promise<'net_on' | 'net_off' | null> {
  const prev = await loadNetState();
  await storeNetState(connected);
  if (prev === null || prev === connected) return null;
  const techId = await loadTechnicianId();
  if (!techId) return null;
  const ev = connected ? 'net_on' : 'net_off';
  await reportDeviceEvent(techId, ev);
  return ev;
}

/**
 * Optimización de batería re-activada / quitada. `optimized = true` significa
 * que el SO la volvió a restringir (el truco clásico para matar el servicio).
 * Solo tiene sentido en Android no-Xiaomi (en Xiaomi la capa MIUI no es legible);
 * el llamador decide si invocarla.
 */
export async function auditBatteryOpt(optimized: boolean): Promise<'battery_restricted' | 'battery_unrestricted' | null> {
  const prev = await loadBattOptState();
  await storeBattOptState(optimized);
  if (prev === null || prev === optimized) return null;
  const techId = await loadTechnicianId();
  if (!techId) return null;
  const ev = optimized ? 'battery_restricted' : 'battery_unrestricted';
  await reportDeviceEvent(techId, ev);
  return ev;
}

/**
 * Permiso de ubicación. Solo importa la transición FUNCIONAL: perder o recuperar
 * "Permitir siempre" (full), que es lo que habilita el rastreo en segundo plano.
 * Bajar a "Solo en uso" (partial) o revocar (none) sin apagar el GPS es un truco
 * para que el rastreo deje de funcionar pareciendo que "la app no sirve".
 * Cambios entre partial/none no se reportan (ya no había rastreo de fondo).
 */
export async function auditPermission(level: PermLevel): Promise<'perm_revoked' | 'perm_granted' | null> {
  const prev = await loadPermState();
  await storePermState(level);
  if (prev === null || prev === level) return null;
  const techId = await loadTechnicianId();
  if (!techId) return null;
  const wasFull = prev === 'full';
  const isFull  = level === 'full';
  if (wasFull && !isFull) { await reportDeviceEvent(techId, 'perm_revoked'); return 'perm_revoked'; }
  if (!wasFull && isFull) { await reportDeviceEvent(techId, 'perm_granted'); return 'perm_granted'; }
  return null;
}

// Hueco mínimo sin fixes para considerar que el rastreo se INTERRUMPIÓ (no un
// simple reinicio rápido de la app). Por debajo de esto no se reporta para no
// generar ruido al reabrir la app a los pocos segundos. Bajado de 5→3 min para
// detectar el force-stop antes (con el heartbeat de fondo, 3 min ya descarta un
// reinicio rápido sin generar falsos positivos en equipos lentos).
const KILLED_GAP_MS = 3 * 60_000;

/**
 * Detecta que una sesión activa quedó sin servicio de ubicación corriendo
 * (force-stop, swipe de recientes, o el SO mató el proceso por memoria) y que
 * hubo un hueco real sin fixes. Se llama al arrancar la app ANTES de revivir el
 * rastreo. Deja un 'tracking_killed' como evidencia de la interrupción.
 *
 * `serviceRunning` = ¿el task de ubicación sigue iniciado? (isTracking()).
 * Devuelve true si registró el evento.
 */
export async function auditTrackingKilled(serviceRunning: boolean): Promise<boolean> {
  if (serviceRunning) return false;
  const techId = await loadTechnicianId();
  if (!techId) return false;                 // no había sesión: arranque limpio
  const lastFix = await loadLastFixTs();
  if (lastFix === null) return false;        // nunca entregó fixes: nada que interrumpir
  if (Date.now() - lastFix < KILLED_GAP_MS) return false;
  await reportDeviceEvent(techId, 'tracking_killed');
  return true;
}

File: android-app/app/src/main/java/com/empresa/localizador/device/DeviceAudit.kt
package com.empresa.localizador.device

import android.content.Context
import com.empresa.localizador.data.Prefs
import com.empresa.localizador.device.EventReporter.DeviceEvent
import com.empresa.localizador.tracking.TrackingConfig
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.ConcurrentHashMap

/**
 * Auditoría de dispositivo: deja evidencia con hora y última posición conocida de
 * las acciones que rompen el rastreo.
 *
 * Detecta TRANSICIONES comparando contra el último estado persistido. Como el
 * estado vive en disco, funciona igual desde la interfaz que desde el servicio o
 * un watchdog con la app cerrada: apagar el GPS o los datos "a escondidas"
 * también deja huella.
 *
 * Reglas comunes:
 *  - Solo se emite si hay sesión de rastreo activa.
 *  - El primer muestreo (estado previo desconocido) solo SIEMBRA, no dispara
 *    evento: abrir la app no debe parecer un sabotaje.
 *  - El estado nuevo se persiste siempre, haya sesión o no, para tener continuidad.
 */
object DeviceAudit {

    private val mutexes = ConcurrentHashMap<String, Mutex>()

    /**
     * Serializa cada auditoría por clave. Sin esto, dos comprobaciones
     * concurrentes de lo mismo (la interfaz sondeando y el servicio a la vez)
     * pueden leer ambas el estado anterior antes de que ninguna guarde el nuevo,
     * detectar las dos la transición y emitir el evento DOS veces.
     */
    private suspend fun <T> locked(key: String, block: suspend () -> T): T =
        mutexes.getOrPut(key) { Mutex() }.withLock { block() }

    fun seed(gpsOn: Boolean? = null, connected: Boolean? = null, perm: PermLevel? = null, batteryOptimized: Boolean? = null) {
        gpsOn?.let { Prefs.gpsState = it }
        connected?.let { Prefs.netState = it }
        perm?.let { Prefs.permState = it.wire }
        batteryOptimized?.let { Prefs.battOptState = it }
    }

    suspend fun auditGps(context: Context, servicesOn: Boolean): DeviceEvent? = locked("gps") {
        val prev = Prefs.gpsState
        Prefs.gpsState = servicesOn
        if (prev == null || prev == servicesOn) return@locked null
        val techId = Prefs.technicianId ?: return@locked null
        val event = if (servicesOn) DeviceEvent.GPS_ON else DeviceEvent.GPS_OFF
        EventReporter.reportDeviceEvent(context, techId, event)
        event
    }

    suspend fun auditNet(context: Context, connected: Boolean): DeviceEvent? = locked("net") {
        val prev = Prefs.netState
        Prefs.netState = connected
        if (prev == null || prev == connected) return@locked null
        val techId = Prefs.technicianId ?: return@locked null
        val event = if (connected) DeviceEvent.NET_ON else DeviceEvent.NET_OFF
        EventReporter.reportDeviceEvent(context, techId, event)
        event
    }

    /**
     * `optimized = true` significa que el sistema volvió a restringir la app: el
     * truco clásico para que el SO mate el servicio.
     */
    suspend fun auditBatteryOpt(context: Context, optimized: Boolean): DeviceEvent? = locked("battopt") {
        val prev = Prefs.battOptState
        Prefs.battOptState = optimized
        if (prev == null || prev == optimized) return@locked null
        val techId = Prefs.technicianId ?: return@locked null
        val event = if (optimized) DeviceEvent.BATTERY_RESTRICTED else DeviceEvent.BATTERY_UNRESTRICTED
        EventReporter.reportDeviceEvent(context, techId, event)
        event
    }

    /**
     * Solo importa la transición FUNCIONAL: perder o recuperar "Permitir siempre",
     * que es lo que habilita el rastreo en segundo plano. Bajarlo a "solo en uso"
     * sin apagar el GPS es la forma silenciosa de romper el rastreo aparentando
     * que "la app no sirve". Los cambios entre parcial y denegado no se reportan:
     * en ambos casos ya no había rastreo de fondo.
     */
    suspend fun auditPermission(context: Context, level: PermLevel): DeviceEvent? = locked("perm") {
        val prev = PermLevel.fromWire(Prefs.permState)
        Prefs.permState = level.wire
        if (prev == null || prev == level) return@locked null
        val techId = Prefs.technicianId ?: return@locked null
        val wasFull = prev == PermLevel.FULL
        val isFull = level == PermLevel.FULL
        when {
            wasFull && !isFull -> {
                EventReporter.reportDeviceEvent(context, techId, DeviceEvent.PERM_REVOKED)
                DeviceEvent.PERM_REVOKED
            }
            !wasFull && isFull -> {
                EventReporter.reportDeviceEvent(context, techId, DeviceEvent.PERM_GRANTED)
                DeviceEvent.PERM_GRANTED
            }
            else -> null
        }
    }

    /**
     * Detecta que una sesión activa se quedó sin servicio corriendo (force-stop,
     * swipe de recientes o muerte por memoria) con un hueco real sin fixes. Se
     * llama al arrancar, ANTES de revivir el rastreo.
     *
     * @return true si registró la interrupción.
     */
    suspend fun auditTrackingKilled(context: Context, serviceRunning: Boolean): Boolean {
        if (serviceRunning) return false
        val techId = Prefs.technicianId ?: return false      // no había sesión
        val lastFix = Prefs.lastFixTs.takeIf { it > 0L } ?: return false  // nunca entregó
        if (System.currentTimeMillis() - lastFix < TrackingConfig.KILLED_GAP_MS) return false
        EventReporter.reportDeviceEvent(context, techId, DeviceEvent.TRACKING_KILLED)
        Prefs.log("watchdog", "El rastreo se había interrumpido; se reanuda")
        return true
    }
}
